include: devicetree: Add port and endpoint macros

ngphibang opened the pull request:
This PR adds some port / endpoint DT helpers for the video drivers or others (e.g. display) and some examples of their _simple_ usage in video drivers. Two main macros helpers are

- `DT_INST_ENDPOINT_BY_ID(inst, pid, id)` : to get the (local) endpoint node id from a port id and an endpoint id
- `DT_NODE_REMOTE_DEVICE(ep)` : to get the peer remote device node from the local endpoint node

_A typical usage would be parsing all endpoints in a loop, or automatically parsing all endpoints till the end._

ngphibang committed:
include: devicetree: Add port and endpoint macro header

Add port and endpoint DT macros to retrieve the node id of the interested
port/endpoint from its id. Also, add helpers to retrieve the peer remote
device node from its local endpoint interface.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>
Co-developed-by: Josuah Demangeon <me@josuah.net>

ngphibang committed:
drivers: video: Use endpoint DT helpers

Drop the driver-defined macros to use the endpoint DT helpers instead.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -202,6 +202,33 @@ struct video_frmival_enum {
+#define DT_REMOTE_DEVICE_NO_PORTS(ep)                                                              \
+	DT_GPARENT(DT_NODELABEL(DT_STRING_TOKEN(ep, remote_endpoint_label)))
+
+#define DEVICE_DT_GET_REMOTE_DEVICE(ep)                                                            \
+	COND_CODE_1(DT_NODE_EXISTS(DT_CHILD(DT_PARENT(DT_REMOTE_DEVICE_NO_PORTS(ep)), ports)), \
+(DEVICE_DT_GET(DT_PARENT(DT_REMOTE_DEVICE_NO_PORTS(ep)))), (DEVICE_DT_GET(DT_REMOTE_DEVICE_NO_PORTS(ep))))
```
A macro with `DEVICE_DT_` prefix would likely go to [`<zephyr/device.h>`](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/device.h). This means it could be interesting to reduce the macro that would ultimately go to `<zephyr/devicetree.h>` down to the DT manipulation:

```suggestion
#define DT_REMOTE_DEVICE(ep)                                                            \
	COND_CODE_1(DT_NODE_EXISTS(DT_CHILD(DT_PARENT(DT_REMOTE_DEVICE_NO_PORTS(ep)), ports)), \
(DT_PARENT(DT_REMOTE_DEVICE_NO_PORTS(ep))), (DT_REMOTE_DEVICE_NO_PORTS(ep)))
```

josuah replied:
Then, `DEVICE_DT_GET_REMOTE_DEVICE(ep)` using it (fated to land in `<zephyr/device.h>`) can be defined like this:

```c
#define DEVICE_DT_GET_REMOTE_DEVICE(ep) DEVICE_DT_GET(DT_REMOTE_DEVICE(ep))
```

josuah replied:
EDIT: I changed my mind: no added value in the suggestion I proposed.

josuah reviewed (changes requested):
It looks typical to have both [`DT_XXX(inst)` and `DT_INST_XXX(inst)` variants](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/devicetree.h#L3815) defined, but not sure if the non-`INST_` would ever be used.
```c
#define DT_INST_XXX(inst) DT_XXX(DT_DRV_INST(inst))
```

Glad to offer the commit/review cycles it takes to get this ready for `<zephyr/drivers/video.h>` or `<zephyr/devicetree.h>`.

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -202,6 +202,33 @@ struct video_frmival_enum {
+	COND_CODE_1(DT_NODE_EXISTS(DT_INST_CHILD(n, ports)), (DT_CHILD(DT_INST_CHILD(n, ports), port_##pid)), (DT_INST_CHILD(n, port_##pid)))
+
+/* Drivers rarely use this as they usually deals with only endpoint */
+#define DT_INST_PORT_BY_ID(n, pid)                                                                 \
+	COND_CODE_1(DT_NODE_EXISTS(_DT_INST_PORT_BY_ID(n, pid)), (_DT_INST_PORT_BY_ID(n, pid)), (DT_INST_CHILD(n, port)))
```
I think it is possible to combine the two previous function in just one, as it is not permitted to have multiple ports (i.e. `port@0`, `port@1`, `port@2`...) without a `ports { ... };` node.

```suggestion
#define DT_INST_PORT_BY_ID(node, pid)                                   \
        COND_CODE_1(DT_NODE_HAS_CHILD(DT_INST_CHILD(node, ports)),      \
                    (DT_CHILD(DT_INST_CHILD(node, ports), port_##pid)), \
                    (DT_INST_CHILD(node, port)))

```

josuah commented:
I tried to summarize my feedback into something like this, which I hope still matches what you had in mind:

```c
/* Handle the variability of "ports{port@0{}};" vs "port{};" while going down */
#define DT_INST_PORT_BY_ID(inst, pid)                                                              \
	COND_CODE_1(DT_NODE_EXISTS(DT_INST_CHILD(inst, ports)),                                    \
		    (DT_CHILD(DT_INST_CHILD(inst, ports), port_##pid)),                            \
		    (DT_INST_CHILD(inst, port)))

/* Handle the variability of "endpoint@0{};" vs "endpoint{};" while going down */
#define DT_INST_ENDPOINT_BY_ID(inst, pid, eid)                                                     \
	COND_CODE_1(DT_NODE_EXISTS(DT_CHILD(DT_INST_PORT_BY_ID(inst, pid), endpoint)),             \
		    (DT_CHILD(DT_INST_PORT_BY_ID(inst, pid), endpoint)),                           \
		    (DT_CHILD(DT_INST_PORT_BY_ID(inst, pid), endpoint_##eid)))

/* Handle the variability of "ports{port@0{}};" vs "port{};" while going up */
#define DT_ENDPOINT_PARENT_DEVICE(node)                                                            \
	COND_CODE_1(DT_NODE_EXISTS(DT_CHILD(DT_GPARENT(node), port)),                              \
		    (DT_GPARENT(node)), (DT_PARENT(DT_GPARENT(node))))

/* Handle the "remote-endpoint-label" */
#define DEVICE_DT_GET_REMOTE_DEVICE(node)                                                          \
	DEVICE_DT_GET(DT_ENDPOINT_PARENT_DEVICE(                                                   \
		DT_NODELABEL(DT_STRING_TOKEN(node, remote_endpoint_label))))
```

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -202,6 +202,33 @@ struct video_frmival_enum {
+	COND_CODE_1(DT_NODE_EXISTS(_DT_INST_PORT_BY_ID(n, pid)), (_DT_INST_PORT_BY_ID(n, pid)), (DT_INST_CHILD(n, port)))
+
+/* Drivers should not use this except they know exactly that the endpoint has a pid */
+#define _DT_INST_ENDPOINT_BY_ID(n, pid, id) DT_CHILD(DT_INST_PORT_BY_ID(n, pid), endpoint_##id)
```
As seen below, it might be possible to have a solution without intermediate macros needed.

ngphibang commented:
> ```
> /* Handle the variability of "ports{port@0{}};" vs "port{};" while going down */
> #define DT_INST_PORT_BY_ID(inst, pid)                                                              \
> 	COND_CODE_1(DT_NODE_EXISTS(DT_INST_CHILD(inst, ports)),                                    \
> 		    (DT_CHILD(DT_INST_CHILD(inst, ports), port_##pid)),                            \
> 		    (DT_INST_CHILD(inst, port)))
> ```

=> This does **not work** when there is no ports node but we have multiple ports, each has a pid but they are not grouped under a ports node. Note that the ports node is **optional**.

That's why I defined two macros so that it **works for all cases**.

```
/* Drivers should not use this except they know exactly that port has a pid */
#define _DT_INST_PORT_BY_ID(n, pid)                                                                \
	COND_CODE_1(DT_NODE_EXISTS(DT_INST_CHILD(n, ports)), (DT_CHILD(DT_INST_CHILD(n, ports), port_##pid)), (DT_INST_CHILD(n, port_##pid)))

#define DT_INST_PORT_BY_ID(n, pid)                                                                 \
	COND_CODE_1(DT_NODE_EXISTS(_DT_INST_PORT_BY_ID(n, pid)), (_DT_INST_PORT_BY_ID(n, pid)), (DT_INST_CHILD(n, port)))
```

Note that the 1st macro is not only an intermediate / internal macro. It can be used by drivers as well if they know exactly that port has a pid.

Same comment for the endpoint macros.

josuah commented:
> => This does **not work** when there is no `ports` node but we still have multiple ports, each has a pid (they are not grouped under a `ports` node though). Note that the `ports` node is **optional**.

I had no idea! This all makes sense then.

josuah reviewed (dismissed):
All good for me! Thank you for your explanations on this.
+1 assuming that:

- Indentation is tuned to get check_compliance.py happy
- Documentation is added (glad to contribute it, but I do not want to break your workflow)
- The naming matches what devicetree maintainers want
- Their introduction in `video.h` is ok as per devicetree maintainers

Should the `devicetree` tag be added?

ngphibang commented:
Relating to the remote endpoint and remote device macros:

```

> /* Handle the variability of "ports{port@0{}};" vs "port{};" while going up */
> #define DT_ENDPOINT_PARENT_DEVICE(node)                                                            \
> 	COND_CODE_1(DT_NODE_EXISTS(DT_CHILD(DT_GPARENT(node), port)),                              \
> 		    (DT_GPARENT(node)), (DT_PARENT(DT_GPARENT(node))))
> 
> /* Handle the "remote-endpoint-label" */
> #define DEVICE_DT_GET_REMOTE_DEVICE(node)                                                          \
> 	DEVICE_DT_GET(DT_ENDPOINT_PARENT_DEVICE(                                                   \
> 		DT_NODELABEL(DT_STRING_TOKEN(node, remote_endpoint_label))))

```

=> You check whether or not we have a `port` node (with no id) while we need to check rather on `ports` node because this will also break when we have multiple ports (with id) but no ports node.

However, by moving the node exist check to the local endpoint of the remote device, this version is simpler / better than mine. Will change to this. Thanks !

ngphibang commented:
> - Their introduction in video.h is ok as per devicetree maintainers.

This is to be discussed. Ideally, it should be in `devicetree.h` / `device.h`. But `remote-endpoint-label` is not recognized system wide except in `video-interfaces` binding. Although we can declare this property in any dts without breaking the build but we cannot use the macros in code without having the corresponding binding. So, to introduce these macros in `devicetree.h` / `device.h`, we need to have also `remote-endpoint-label` in Zephyr generic bindings (`base.yaml` ?). How ?

If we keep them only for video, it's ok for the 1st phase. But do you know if there are examples in Zephyr that subsystem can have  their own DT macros ? or all DT macros have to go to `devicetree.h` ?

josuah commented:
It looks like driver classes *do* have their own macros! In a subdirectory of [`include/zephyr/devicetree`](https://github.com/zephyrproject-rtos/zephyr/tree/main/include/zephyr/devicetree).

For instance `include/zephyr/devicetree/video.h` file with macro named like `DT_VIDEO_...` or `DT_INST_VIDEO_...` if following the same pattern.

https://github.com/zephyrproject-rtos/zephyr/blob/4bb945300bf92e4708404102d2a3b54e540bfd6c/include/zephyr/devicetree.h#L5146-L5158

When the macros generate a struct specific to one subsystem (i.e. `i2c_dt_spec`), they are put in the `include/zephyr/drivers/<class>.h` with a name typically like `<CLASS>_DT_...`:
https://github.com/zephyrproject-rtos/zephyr/blob/4bb945300bf92e4708404102d2a3b54e540bfd6c/include/zephyr/drivers/i2c.h#L92-L94

danieldegrasse commented on include/zephyr/drivers/video.h:
```diff
@@ -202,6 +202,33 @@ struct video_frmival_enum {
+ */
+
+/* Drivers should not use this except they know exactly that port has a pid */
+#define _DT_INST_PORT_BY_ID(n, pid)                                                                \
+	COND_CODE_1(DT_NODE_EXISTS(DT_INST_CHILD(n, ports)), (DT_CHILD(DT_INST_CHILD(n, ports), port_##pid)), (DT_INST_CHILD(n, port_##pid)))
```
Sorry if I'm missing something here, why not use this pattern? Or remove the underscore if users should be using this macro directly?

ngphibang replied:
We can use this pattern directly if we are sure that the target port has an ID, e.g. port_0, port_1, etc. It will not work in case we have only 1 port with no ID. I will remove the underscore as this macro can be used both internally and publicly.

danieldegrasse commented on include/zephyr/drivers/video.h:
```diff
@@ -202,6 +202,33 @@ struct video_frmival_enum {
+	COND_CODE_1(DT_NODE_EXISTS(_DT_INST_ENDPOINT_BY_ID(n, pid, id)), (_DT_INST_ENDPOINT_BY_ID(n, pid, id)), (DT_CHILD(DT_INST_PORT_BY_ID(n, pid), endpoint)))
+
+/* Drivers should not use this except they know exactly that the remote device has no ports node */
+#define DT_REMOTE_DEVICE_NO_PORTS(ep)                                                              \
```
Question here- it seems like in Zephyr we only currently support this scheme for remote ports:
```
	mt9m114: mt9m114@48 {
		compatible = "aptina,mt9m114";
		reg = <0x48>;

		port {
			mt9m114_ep_out: endpoint {
				remote-endpoint-label = "dfi_ep_in";
			};
		};
	};
```

So do we need to supply this macro yet?

danieldegrasse replied:
Ok, I see how we are doing this with multiple child bindings now- sorry I missed that.

danieldegrasse reviewed (commented):
General comment- we need additional documentation on these macros before we can move forwards with this PR, I assume that will be added?

ngphibang commented:
> General comment- we need additional documentation on these macros before we can move forwards with this PR, I assume that will be added?

Yes, I will put documentation and other things as discussed in the next push.

ngphibang commented:
force-push

- Some minor fix
- Fix compliance check
- Simplify / Optimize `DEVICE_DT_GET_REMOTE_DEVICE`  a bit. It takes me quite some time to recall what I have planned to optimize this macro so I ran out of time to add documentation.

@josuah If you want to help to put some documentation then go ahead :-) otherwise I will do it later.

josuah commented:
> @josuah If you want to help to put some documentation then go ahead :-) otherwise I will do it later.

Yes! Docs and tests, I will take inspiration from the other PR.

josuah commented:
https://github.com/josuah/zephyr/tree/add_endpoint_parser_doc
- [x] documentation
- [ ] tests.

josuah committed:
tests: lib: devicetree: Add tests for endpoint DT macros

Add tests for the new port / endpoint DT macros

Signed-off-by: Josuah Demangeon <me@josuah.net>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

josuah commented:
I think it is ready to go.

`west twister --inline-logs --platform native_sim --scenario tests/lib/devicetree/api/libraries.devicetree.api`
<details><summary>output</summary><pre>
Renaming output directory to /home/tinyvision/zephyrproject/zephyr/twister-out.31
INFO    - Using Ninja..
INFO    - Zephyr version: v4.0.0-1159-g6a028d0c99a7
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/tinyvision/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 1
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:    1/   1  100%  built (not run):    0, filtered:    0, failed:    0, error:    0
INFO    - 1 test scenarios (1 configurations) selected, 0 configurations filtered (0 by static filter, 0 at runtime).
INFO    - 1 of 1 executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 21.96 seconds.
INFO    - 71 of 71 executed test cases passed (100.00%) on 1 out of total 871 platforms (0.11%).
INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/tinyvision/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/tinyvision/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/tinyvision/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
</pre></details>

ngphibang commented:
@josuah : Thank you for the nice work ! I am going to squash your documentation (with just some minor modifications) commit and cherry-pick the test commit now.

- https://github.com/zephyrproject-rtos/zephyr/commit/e23305727c467b51a99d43816c4271080fe8c655#diff-d3ce6b001979843da3e07890fd4bdbf9008ee4fb2c813c95baf108be1ec2b539R1323 and same for line 1325
=> `DT_INST_ENDPOINT_BY_ID(1, 0, 0)` should be `DT_INST_ENDPOINT_BY_ID(0, 0, 0)` ?

- For `DEVICE_DT_GET_REMOTE_DEVICE` tests, I will change the text to "get **remote** device of ..." as it gets the remote device.

ngphibang commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -1309,6 +1324,78 @@ ZTEST(devicetree_api, test_dma)
+			     POST_KERNEL, CONFIG_APPLICATION_INIT_PRIORITY, NULL);
+ZTEST(devicetree_api, test_video)
+{
+	/* DT_SAME_NODE */
```
`/* DT_INST_PORT_BY_ID */`

ngphibang commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -1309,6 +1324,78 @@ ZTEST(devicetree_api, test_dma)
+	/* DT_SAME_NODE */
+#define DT_DRV_COMPAT vnd_video_single_port
+	zassert_true(DT_SAME_NODE(DT_INST_PORT_BY_ID(0, 0),
+				  TEST_VIDEO0_PORT), "get video 0 port 0");
```
`"get port0 node of the video0 device")`
same for others

josuah commented:
> => DT_INST_ENDPOINT_BY_ID(1, 0, 0) should be DT_INST_ENDPOINT_BY_ID(0, 0, 0) ?
> => DT_INST_PORT_BY_ID(1, 0) => DT_INST_PORT_BY_ID(0, 0) ?

Here the naming is intuitive:

```c
zassert_true(DT_SAME_NODE(DT_INST_ENDPOINT_BY_ID(0, 0, 0),
                                    ,------------'
                          TEST_VIDEO0_OUT), "get video 0 port 0 endpoint 0");

zassert_true(DT_SAME_NODE(DT_INST_ENDPOINT_BY_ID(1, 0, 0),
                                    ,------------'
                          TEST_VIDEO1_OUT0), "get video 1 port 0 endpoint 0");

zassert_true(DT_SAME_NODE(DT_INST_ENDPOINT_BY_ID(1, 0, 1),
                                    ,------------'
                          TEST_VIDEO1_OUT1), "get video 1 port 0 endpoint 1");
zassert_true(DT_SAME_NODE(DT_INST_ENDPOINT_BY_ID(0, 0, 0),
                                    ,------------'
                          TEST_VIDEO0_OUT), "get video 0 port 0 endpoint 0");

zassert_true(DT_SAME_NODE(DT_INST_ENDPOINT_BY_ID(1, 0, 0),
                                    ,------------'
                          TEST_VIDEO1_OUT0), "get video 1 port 0 endpoint 0");

zassert_true(DT_SAME_NODE(DT_INST_ENDPOINT_BY_ID(1, 0, 1),
                                    ,------------'
                          TEST_VIDEO1_OUT1), "get video 1 port 0 endpoint 1");
```

Here the naming is counter-intuitive:
```c
zassert_true(DT_SAME_NODE(DT_INST_ENDPOINT_BY_ID(0, 0, 0),
                                    .------------'
                          TEST_VIDEO2_PORT0_IN0), "get video 2 port 0 endpoint 0");

zassert_true(DT_SAME_NODE(DT_INST_ENDPOINT_BY_ID(0, 0, 1),
                                    .------------'
                          TEST_VIDEO2_PORT0_IN1), "get video 2 port 0 endpoint 1");

zassert_true(DT_SAME_NODE(DT_INST_ENDPOINT_BY_ID(0, 1, 0),
                                    .------------'
                          TEST_VIDEO2_PORT1_IN), "get video 2 port 1 endpoint 0");
```

It is because `VIDEO2` is a *different* `compatble = "";`...

Maybe `VIDEOSINGLE0`, `VIDEOSINGLE1`, `VIDEOMULTI0` would be better names than `VIDEO0`, `VIDEO1`, **`VIDEO2`** (the last one having the same name, but different compatible).

I could not find a way to have bindings supporting both `ports { port { endpoint {}; }; };` and `port { endpoint {}; };`, so I went with two separate bindings.

ngphibang commented:
Ok, got it now ... I didn't see that `video0` and `video1` share the same `compatible` so they will be on different instances. Will reput things as before then.

ngphibang commented:
I think it's ready now. Could you revisit this PR @josuah @danieldegrasse ? Thanks !

danieldegrasse commented on include/zephyr/drivers/video.h:
```diff
@@ -202,6 +202,220 @@ struct video_frmival_enum {
+	COND_CODE_1(DT_NODE_EXISTS(_DT_INST_PORT_BY_ID(inst, pid)),                                \
+		(_DT_INST_PORT_BY_ID(inst, pid)), (DT_INST_CHILD(inst, port)))
+
+#define _DT_INST_ENDPOINT_BY_ID(inst, pid, eid)                                                    \
```
I *think* these still need to be documented, right? Even if the documentation is simply "Helper for DT_INST_PORT_BY_ID"

ngphibang replied:
done

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -202,6 +202,220 @@ struct video_frmival_enum {
+ * @code{.dts}
+ *	&device {
+ *		ports {
+ *			port@0 {
```
Note: I did not include the `#size-cells = <0>;` and `#address-cells = <1>` here for compactness.
No idea if it is best to add that or not.

ngphibang replied:
yes, I think we should. Whenever there is multiple ports or endpoints, the parent node need to declare `#size-cells` and `#address-cells`

ngphibang replied:
done

danieldegrasse commented on include/zephyr/drivers/video.h:
Should we place these definitions in `include/zephyr/devicetree/video.h`? I also question placing these in `video.h` in general- these will probably be useful for displays too, right?

ngphibang replied:
> I also question placing these in video.h in general- these will probably be useful for displays too, right?

I think you mean placing these in `devicetree.h` and `device.h` ? Yes, we wanted to do so. But as being mentionned in the PR's description, these macros are based on the `remote-endpoint-label`property that is only defined in the `video-interfaces.yaml `and which is specific to video and not recognized system-wide.

Is it ok to put these in `devicetree.h` even if the `remote-endpoint-label` is not defined for other subsystem e.g. display ? then let each subsystem define them or there is a way to define this property system-wide ?

danieldegrasse replied:
I'll leave it to @decsny if we should place these in `devicetree.h` or not. My suggestion was `devicetree/video.h`, to follow the pattern used for other helpers like GPIO: https://github.com/zephyrproject-rtos/zephyr/blob/e60f04096cd8d9227c8279ca830e72bc1b5b8b1c/include/zephyr/devicetree/gpio.h#L1 Those helpers are specific to GPIO, and the devicetree properties they access wouldn't be available to all subsystems.

IMO, I'd like to place these in a header like `include/zephyr/devicetree/endpoints.h`, but I'm open to other options.

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -202,6 +202,220 @@ struct video_frmival_enum {
+ * @ref DT_INST_PORT_BY_ID and @ref DT_INST_ENDPOINT_BY_ID.
+ *
+ * @code{.c}
+ *	const struct device *dev1 = DEVICE_DT_GET_REMOTE_DEVICE(DT_NODELABEL(ep0))
```
There is no `ep0:` nodelabel in the snippet below. Maybe the example below or this snippet can be updated?

ngphibang replied:
yes, it should be `device0_ep_out`, will change.

ngphibang replied:
done

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -202,6 +202,220 @@ struct video_frmival_enum {
+ * @endcode
+ *
+ * @param ep endpoint node
+ * @return device node of @n
```
I am learning Doxygen and [Breathe](https://breathe.readthedocs.io/en/latest/index.html) as I go, but `@n` is the equivalent of `\n`, and this [introduces a newline](https://www.doxygen.nl/manual/commands.html#cmdn). Maybe [`@p ep`](https://www.doxygen.nl/manual/commands.html#cmdp) or [`@c n`](https://www.doxygen.nl/manual/commands.html#cmdc) was meant?

ngphibang replied:
I don't know neither :-). I didn't find any place in Zephyr that use `@n` but forgot to ask you. So, what I am supposed to place here to replace `@n` ?

josuah replied:
I think something like this would work better: `@p` for parameter. There is no `@p n` parameter here.
```suggestion
 * @return device node associated with @p ep
```

ngphibang replied:
done

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -202,6 +202,220 @@ struct video_frmival_enum {
+ * @endcode
+ *
+ * @param ep endpoint node
+ * @return pointer to the remote device object that connects to this device via @n
```
Same as above for `@n`

josuah commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -1309,6 +1324,78 @@ ZTEST(devicetree_api, test_dma)
+#undef DT_DRV_COMPAT
+#define DT_DRV_COMPAT vnd_video_multi_port
+	zassert_true(DT_SAME_NODE(DT_INST_ENDPOINT_BY_ID(0, 0, 0), TEST_VIDEO2_PORT0_IN0),
+		     "get endpoint@0 node of port@0 node of video2");
```
This is a much better help text, thank you.

josuah reviewed (changes requested):
I think I left some typo around use of `@n`.
Some other non-blocking comments.

rruuaanng commented on dts/bindings/test/vnd,video-multi-port.yaml:
```diff
@@ -0,0 +1,14 @@
+      include: video-interfaces.yaml
+      properties:
+        reg:
+          type: array
```
Add a newline to the end of the file
```suggestion
          type: array

```

ngphibang replied:
Will do, thanks !

ngphibang replied:
done

rruuaanng commented on dts/bindings/test/vnd,video-single-port.yaml:
```diff
@@ -0,0 +1,13 @@
+    include: video-interfaces.yaml
+    properties:
+      reg:
+        type: array
```
+1

rruuaanng commented on include/zephyr/devicetree/port-endpoint.h:
```diff
@@ -0,0 +1,267 @@
+ *
+ * Given a device instance number, a port ID and an endpoint ID, return the endpoint node.
+ * It handles various ways of how a port and an endpoint could be defined as described in
+ * @ref DT_INST_PORT_BY_ID and below. For example:
```
It's recommended to replace `For example` with this
```
 * Example devicetree fragment:
```
`devicetree.h` currently almost all use this, and I don’t want to see their inconsistency, this is necessary for the neatness of the file.

ngphibang replied:
Will do, thanks !

rruuaanng commented on include/zephyr/devicetree/port-endpoint.h:
```diff
@@ -0,0 +1,267 @@
+ *
+ * Given an endpoint node id, return its device node id.
+ * This handles various ways of how a port and an endpoint could be defined as described in
+ * @ref DT_NODE_BY_ENDPOINT. For example:
```
+1

rruuaanng commented on include/zephyr/devicetree/port-endpoint.h:
```diff
@@ -0,0 +1,267 @@
+ *
+ * Given an endpoint node id, return the remote device node that connects to this device via this
+ * local endpoint. This handles various ways of how a port and an endpoint could be defined as
+ * described in @ref DT_INST_PORT_BY_ID and @ref DT_INST_ENDPOINT_BY_ID. For example:
```
+1

rruuaanng commented on include/zephyr/devicetree/port-endpoint.h:
```diff
@@ -0,0 +1,267 @@
+ * @brief Get a port node from its id
+ *
+ * Given a device instance number, return a port node specified by its ID.
+ * It handles various ways of how a port could be defined. For example:
```
+1

ngphibang commented:
force-push:

- Resolved current discussions
- Place these macros under `include/zephyr/devicetree/port-endpoint.h` which effectively included in `devicetree.h`.
- Changed the last macro DEVICE_DT_GET_REMOTE_DEVICE(ep) to DT_NODE_REMOTE_DEVICE(ep) because all the macros should be devicetree related only so that we don't have to move this to device.h. Doing so will make us lost a level of abstraction but it's not a big deal. I changed the test cases to test this new macro too.
- Fix some minor errors and wording (remove the `property` examples).

danieldegrasse reviewed (dismissed):
Looks fine to me, thank you for this update

ngphibang commented:
> Looks fine to me, thank you for this update

I still need to resolve @rruuaanng 's comments, will do soon :-).

ngphibang commented:
@rruuaanng : I tried to add blank lines to the files as your request but then check-compliance has failed:

```
ERROR   : Test YAMLLint failed: 
YAMLLint (empty-lines):too many blank lines (1 > 0)
File:dts/bindings/test/vnd,video-multi-port.yaml
Line:15
Column:1
ERROR   : Test YAMLLint failed: 
YAMLLint (empty-lines):too many blank lines (1 > 0)
File:dts/bindings/test/vnd,video-single-port.yaml
Line:14
Column:1
ERROR   : Test Nits failed: 
Please remove blank lines at end of 'dts/bindings/test/vnd,video-multi-port.yaml'
ERROR   : Test Nits failed: 
Please remove blank lines at end of 'dts/bindings/test/vnd,video-single-port.yaml'
ERROR   : Test GitDiffCheck failed: 
0e805944c8829c236348b4dcdd41e4f9fb8eb6a0: dts/bindings/test/vnd,video-multi-port.yaml:15: new blank line at EOF.
0e805944c8829c236348b4dcdd41e4f9fb8eb6a0: dts/bindings/test/vnd,video-single-port.yaml:14: new blank line at EOF.
```
In fact, I see **there is already a blank line at the end of the file in my local IDE but github does not show it**.
So, I will let it as before then ?

ngphibang commented:
@rruuaanng No problem :-). Done.

ngphibang commented:
force-push to fix CI build issue due to https://github.com/zephyrproject-rtos/zephyr/pull/79482 (which take these macros inside it) get merged first

ngphibang commented:
@danieldegrasse Could you revisit ?

danieldegrasse reviewed (approved):
@decsny could you take a look as well?
